File: verify_lawrence_conjugacy_B4.py
#!/usr/bin/env python3
"""
verify_lawrence_conjugacy_B4.py

Verifies that the iterated reduced Long-Moody representation of B_4
matches Lawrence's representation by checking:
1. Both satisfy B_4 braid relations
2. Corresponding generators have the same eigenvalues (characteristic polynomials)
"""

from sympy import symbols, eye, zeros, Matrix, cancel, factor, Poly, Symbol

Q, q, a = symbols('Q q a')
lam = Symbol('lambda')


def build_Ni(i, gi, n, d):
    nb = n - 1
    Id = eye(d)
    Zd = zeros(d)
    blocks = [[Id if r == c else Zd for c in range(nb)] for r in range(nb)]
    if i == 0:
        blocks[0][0] = -gi; blocks[0][1] = Zd
        blocks[1][0] = Id;  blocks[1][1] = Id
    elif i == n - 2:
        blocks[nb-2][nb-2] = Id;  blocks[nb-2][nb-1] = gi
        blocks[nb-1][nb-2] = Zd;  blocks[nb-1][nb-1] = -gi
    else:
        p = i - 1
        blocks[p][p] = Id;    blocks[p][p+1] = gi;    blocks[p][p+2] = Zd
        blocks[p+1][p] = Zd;  blocks[p+1][p+1] = -gi; blocks[p+1][p+2] = Zd
        blocks[p+2][p] = Zd;  blocks[p+2][p+1] = Id;  blocks[p+2][p+2] = Id
    return blocks


def left_multiply_blocks(blocks, M, nb):
    return [[M * blocks[r][c] for c in range(nb)] for r in range(nb)]


def multiply_block_matrices(A, B, nb, d):
    C = [[zeros(d) for _ in range(nb)] for _ in range(nb)]
    for r in range(nb):
        for c in range(nb):
            acc = zeros(d)
            for k in range(nb):
                acc = acc + A[r][k] * B[k][c]
            C[r][c] = acc
    return C


def assemble(blocks, nb, d):
    N = nb * d
    M = zeros(N)
    for r in range(nb):
        for c in range(nb):
            M[r*d:(r+1)*d, c*d:(c+1)*d] = blocks[r][c]
    return M


def reduced_lm_colored(rho_g, rho_sigma, partition=None):
    n = len(rho_g)
    num_sigmas = len(rho_sigma)
    d = rho_g[0].shape[0]
    if partition is None:
        partition = [1] * n
    is_boundary = [partition[i] != partition[i+1] for i in range(num_sigmas)]
    nb = n - 1
    rho_plus = []
    for i in range(num_sigmas):
        if not is_boundary[i]:
            blocks = build_Ni(i, rho_g[i], n, d)
            blocks = left_multiply_blocks(blocks, rho_sigma[i], nb)
            rho_plus.append(cancel(assemble(blocks, nb, d)))
        else:
            gi_conj = cancel(rho_g[i] * rho_g[i+1] * rho_g[i].inv())
            blocks_N  = build_Ni(i, rho_g[i], n, d)
            blocks_Np = build_Ni(i, gi_conj, n, d)
            blocks_prod = multiply_block_matrices(blocks_Np, blocks_N, nb, d)
            si_sq = rho_sigma[i] * rho_sigma[i]
            blocks_prod = left_multiply_blocks(blocks_prod, si_sq, nb)
            rho_plus.append(cancel(assemble(blocks_prod, nb, d)))
    return rho_plus


def free_group_gens(rho_sigma, partition=None):
    n = len(rho_sigma)
    if partition is None:
        partition = [1] * (n + 1)
    last_is_boundary = (partition[n-1] != partition[n])
    if last_is_boundary:
        gn = rho_sigma[n-1]
    else:
        gn = rho_sigma[n-1] * rho_sigma[n-1]
    rho_g = [None] * n
    rho_g[n-1] = cancel(gn)
    for i in range(n-2, -1, -1):
        rho_g[i] = cancel(rho_sigma[i].inv() * rho_g[i+1] * rho_sigma[i])
    return rho_g


def lawrence_generators(n, i, q1, q2, alpha):
    d = n * (n - 1)
    M = zeros(d)
    pairs = []
    idx = {}
    count = 0
    for la in range(1, n+1):
        for mu in range(1, n+1):
            if la != mu:
                pairs.append((la, mu))
                idx[(la, mu)] = count
                count += 1
    def p(a, b): return idx[(a, b)]
    si = i + 1; si1 = si + 1

    r = p(si, si1)
    M[r, p(si1, si)] = 1/(q1*q2)
    for j in range(1, si):
        M[r, p(si1, j)] = (1/q2 - 1)/q1
    for k in range(si1+1, n+1):
        M[r, p(si1, k)] = (1/q2 - 1)/q1

    r = p(si1, si)
    M[r, p(si, si1)] = 1/(q1*q2*alpha)
    for j in range(1, si):
        M[r, p(j, si1)] = (1/q1 - 1)/(q2*alpha)
    for k in range(si1+1, n+1):
        M[r, p(k, si1)] = (1/q1 - 1)/q2

    for j in range(1, si):
        M[p(si,j), p(si1,j)] = 1/q1
        M[p(j,si), p(j,si1)] = 1/q2
        M[p(si1,j), p(si,j)] = 1; M[p(si1,j), p(si1,j)] = 1-1/q1
        M[p(j,si1), p(j,si)] = 1; M[p(j,si1), p(j,si1)] = 1-1/q2

    for k in range(si1+1, n+1):
        M[p(si,k), p(si1,k)] = 1/q1
        M[p(k,si), p(k,si1)] = 1/q2
        M[p(si1,k), p(si,k)] = 1; M[p(si1,k), p(si1,k)] = 1-1/q1
        M[p(k,si1), p(k,si)] = 1; M[p(k,si1), p(k,si1)] = 1-1/q2

    for s_idx in range(d):
        la, mu = pairs[s_idx]
        if la != si and la != si1 and mu != si and mu != si1:
            M[s_idx, s_idx] = 1
    return M


def check_zero(M):
    return all(cancel(M[i, j]) == 0 for i in range(M.rows) for j in range(M.cols))


def charpoly(M):
    """Characteristic polynomial, factored."""
    return factor(M.charpoly(lam).as_expr())


if __name__ == '__main__':
    # === Step 1 ===
    print('=== Step 1: First reduced Long-Moody iteration ===')
    rho_g_seed = [Matrix([[Q]]), Matrix([[Q]]), Matrix([[Q]]), Matrix([[Q]]), Matrix([[a]])]
    rho_s_seed = [Matrix([[1]]), Matrix([[1]]), Matrix([[1]]), Matrix([[1]])]
    L4Pre = reduced_lm_colored(rho_g_seed, rho_s_seed, [1,1,1,1,2])
    print(f'  Output: {len(L4Pre)} matrices, each {L4Pre[0].shape[0]}x{L4Pre[0].shape[1]}')

    # === Step 2 ===
    print('\n=== Step 2: Recover free group generators ===')
    L4PreFree = free_group_gens(L4Pre, [1,1,1,1,2])
    print(f'  Recovered g_1,...,g_4 as {L4PreFree[0].shape[0]}x{L4PreFree[0].shape[1]} matrices')

    # === Step 3 ===
    print('\n=== Step 3: Second reduced Long-Moody iteration ===')
    rho_g2 = [g * q for g in L4PreFree]
    L4 = reduced_lm_colored(rho_g2, L4Pre[0:3])
    s1, s2, s3 = L4[0], L4[1], L4[2]
    print(f'  Output: {len(L4)} matrices, each {s1.shape[0]}x{s1.shape[1]}')

    # === Step 4 ===
    print('\n=== Step 4: Compute Lawrence representation for n=4 ===')
    L_s1 = lawrence_generators(4, 0, 1/Q, 1/q, 1/a)
    L_s2 = lawrence_generators(4, 1, 1/Q, 1/q, 1/a)
    L_s3 = lawrence_generators(4, 2, 1/Q, 1/q, 1/a)
    print(f'  Lawrence matrices: each {L_s1.shape[0]}x{L_s1.shape[1]}')

    # === Step 5: Braid relations ===
    print('\n=== Step 5: Braid relations ===')
    print('  Our representation:')
    print(f'    s1*s2*s1 = s2*s1*s2: {check_zero(cancel(s1*s2*s1 - s2*s1*s2))}')
    print(f'    s2*s3*s2 = s3*s2*s3: {check_zero(cancel(s2*s3*s2 - s3*s2*s3))}')
    print(f'    s1*s3 = s3*s1:       {check_zero(cancel(s1*s3 - s3*s1))}')

    print('  Lawrence representation:')
    print(f'    s1*s2*s1 = s2*s1*s2: {check_zero(cancel(L_s1*L_s2*L_s1 - L_s2*L_s1*L_s2))}')
    print(f'    s2*s3*s2 = s3*s2*s3: {check_zero(cancel(L_s2*L_s3*L_s2 - L_s3*L_s2*L_s3))}')
    print(f'    s1*s3 = s3*s1:       {check_zero(cancel(L_s1*L_s3 - L_s3*L_s1))}')

    # === Step 6: Compare characteristic polynomials ===
    print('\n=== Step 6: Compare characteristic polynomials ===')
    for name, ours, law in [('sigma_1', s1, L_s1), ('sigma_2', s2, L_s2), ('sigma_3', s3, L_s3)]:
        cp_ours = charpoly(ours)
        cp_law  = charpoly(law)
        diff = factor(cp_ours - cp_law)
        match = (diff == 0)
        print(f'  {name}: {"MATCH" if match else "MISMATCH"}')
        if not match:
            print(f'    Ours:     {cp_ours}')
            print(f'    Lawrence: {cp_law}')
            print(f'    Diff:     {diff}')

    print('\n=== Done. ===')

File: verify_lawrence_conjugacy_B4.m
% verify_lawrence_conjugacy_B4.m
%
% Proves that the iterated reduced Long-Moody representation of B_4 on
% 2-point ordered configurations is simultaneously conjugate to Lawrence's
% representation from Fig 4.4 of Lawrence (1990).
%
% This extends verify_lawrence_conjugacy.m from n=3 to n=4.
%
% Parameters: Q (z_2 monodromy), q (z_1 monodromy), a (z_1-z_2 monodromy)
% Lawrence params: q1 = 1/Q, q2 = 1/q, alpha = 1/a

syms Q q a;

%% === Step 1: First iteration of reduced Long-Moody ===
%
% Seed representation of F_5 ⋊ B_5 (five strands, last one colored differently):
%   rho_g     = {Q, Q, Q, Q, a}  — 5 scalars (1x1 matrices)
%   rho_sigma = {1, 1, 1, 1}     — 4 scalars, all trivial
%   partition = [1 1 1 1 2]       — strand 5 is a different color
%
% Output: 4 matrices, each 4x4.
%   L4Pre{1} = rho^+(sigma_1)
%   L4Pre{2} = rho^+(sigma_2)
%   L4Pre{3} = rho^+(sigma_3)
%   L4Pre{4} = rho^+(sigma_4^2)  (squared because of partition boundary)

fprintf('=== Step 1: First reduced Long-Moody iteration ===\n');
fprintf('  Seed: rho_g = {Q,Q,Q,Q,a}, rho_sigma = {1,1,1,1}, partition = [1 1 1 1 2]\n');
L4Pre = reduced_lm_colored({Q, Q, Q, Q, a}, {sym(1), sym(1), sym(1), sym(1)}, [1 1 1 1 2]);
fprintf('  Output: 4 matrices, each %dx%d\n', size(L4Pre{1}));

%% === Step 2: Recover free group generators ===
%
% Using Birman exact sequence with partition [1 1 1 1 2]:
%   g_4 = L4Pre{4}            (already sigma_4^2 since last index is boundary)
%   g_3 = sigma_3^{-1} * g_4 * sigma_3  = L4Pre{3}^{-1} * g_4 * L4Pre{3}
%   g_2 = sigma_2^{-1} * g_3 * sigma_2  = L4Pre{2}^{-1} * g_3 * L4Pre{2}
%   g_1 = sigma_1^{-1} * g_2 * sigma_1  = L4Pre{1}^{-1} * g_2 * L4Pre{1}
%
% Output: {g_1, g_2, g_3, g_4} — four 4x4 matrices.

fprintf('\n=== Step 2: Recover free group generators ===\n');
L4PreFree = free_group_gens(L4Pre, [1 1 1 1 2]);
fprintf('  Recovered g_1,...,g_4 as %dx%d matrices\n', size(L4PreFree{1}));

%% === Step 3: Second iteration of reduced Long-Moody ===
%
% Seed representation of F_4 ⋊ B_4:
%   rho_g     = {g_1*q, g_2*q, g_3*q, g_4*q}  — 4 matrices, each 4x4, scaled by q
%   rho_sigma = {L4Pre{1}, L4Pre{2}, L4Pre{3}} — 3 matrices, each 4x4 (braid gens of B_4)
%   partition = [1 1 1 1]                        — uniform (default), no boundaries
%
% Output: 3 matrices, each 12x12.
%   L4{1} = rho^{++}(sigma_1)
%   L4{2} = rho^{++}(sigma_2)
%   L4{3} = rho^{++}(sigma_3)

fprintf('\n=== Step 3: Second reduced Long-Moody iteration ===\n');
fprintf('  Seed: rho_g = {g_i*q}, rho_sigma = {L4Pre{1..3}}, no partition\n');
rho_g = {L4PreFree{1}*q, L4PreFree{2}*q, L4PreFree{3}*q, L4PreFree{4}*q};
L4 = reduced_lm_colored(rho_g, L4Pre(1:3));
s1 = simplify(L4{1});
s2 = simplify(L4{2});
s3 = simplify(L4{3});
fprintf('  Output: 3 matrices, each %dx%d\n', size(s1));

%% === Step 4: Compute Lawrence's representation directly ===
%
% Lawrence (1990) Fig 4.4 for n=4, k=2 (ordered 2-point configurations).
% Basis: n(n-1) = 12 ordered pairs (lambda, mu), lambda ~= mu in {1,2,3,4},
% in lexicographic order:
%   1:(1,2) 2:(1,3) 3:(1,4) 4:(2,1) 5:(2,3) 6:(2,4)
%   7:(3,1) 8:(3,2) 9:(3,4) 10:(4,1) 11:(4,2) 12:(4,3)
%
% Parameter correspondence: q1 = 1/Q, q2 = 1/q, alpha = 1/a

fprintf('\n=== Step 4: Compute Lawrence representation for n=4 ===\n');
fprintf('  Basis: 12 ordered pairs (lam,mu) in lex order\n');
fprintf('  Parameters: q1 = 1/Q, q2 = 1/q, alpha = 1/a\n');
L_s1 = simplify(lawrence_generators(4, 1, 1/Q, 1/q, 1/a));
L_s2 = simplify(lawrence_generators(4, 2, 1/Q, 1/q, 1/a));
L_s3 = simplify(lawrence_generators(4, 3, 1/Q, 1/q, 1/a));
fprintf('  Lawrence matrices: 3 generators, each %dx%d\n', size(L_s1));

%% === Step 5: Find intertwining matrix P ===
%
% Solve for P such that P * s_i = L_s_i * P for all i simultaneously.
% Vectorize: (s_i^T ⊗ I) vec(P) = (I ⊗ L_s_i) vec(P)
% i.e., [(s_i^T ⊗ I) - (I ⊗ L_s_i)] vec(P) = 0 for each i.
% Stack all three constraints and find the null space.

fprintf('\n=== Step 5: Find intertwining matrix P ===\n');
fprintf('  Solving (s_i^T ⊗ I_12 - I_12 ⊗ L_s_i) vec(P) = 0 for i=1,2,3\n');
I12 = sym(eye(12));
A1 = kron(s1.', I12) - kron(I12, L_s1);
A2 = kron(s2.', I12) - kron(I12, L_s2);
A3 = kron(s3.', I12) - kron(I12, L_s3);
N = null([A1; A2; A3]);
fprintf('  Null space dimension: %d\n', size(N, 2));
assert(size(N, 2) == 1, 'Expected 1-dimensional null space (representations are irreducible).');

P = simplify(reshape(N(:,1), [12, 12]));
d = simplify(det(P));
fprintf('  det(P) = '); disp(d);
assert(~isequal(d, sym(0)), 'P must be invertible for generic Q, q, a.');
fprintf('  P is invertible for generic Q, q, a.\n');

%% === Step 6: Verify simultaneous conjugacy ===

fprintf('\n=== Step 6: Verify P * s_i = L_s_i * P for i = 1, 2, 3 ===\n');

check1 = simplify(P * s1 - L_s1 * P);
check2 = simplify(P * s2 - L_s2 * P);
check3 = simplify(P * s3 - L_s3 * P);

pass1 = all(all(check1 == 0));
pass2 = all(all(check2 == 0));
pass3 = all(all(check3 == 0));

if pass1, fprintf('  PASS: P * sigma_1 = L_sigma_1 * P\n');
else,      fprintf('  FAIL: sigma_1 conjugacy\n'); end

if pass2, fprintf('  PASS: P * sigma_2 = L_sigma_2 * P\n');
else,      fprintf('  FAIL: sigma_2 conjugacy\n'); end

if pass3, fprintf('  PASS: P * sigma_3 = L_sigma_3 * P\n');
else,      fprintf('  FAIL: sigma_3 conjugacy\n'); end

%% === Step 7: Verify braid relations for both representations ===
%
% B_4 braid relations:
%   (i)   sigma_1 sigma_2 sigma_1 = sigma_2 sigma_1 sigma_2
%   (ii)  sigma_2 sigma_3 sigma_2 = sigma_3 sigma_2 sigma_3
%   (iii) sigma_1 sigma_3 = sigma_3 sigma_1  (far commutativity)

fprintf('\n=== Step 7: Verify braid relations ===\n');

our_braid12  = simplify(s1*s2*s1 - s2*s1*s2);
our_braid23  = simplify(s2*s3*s2 - s3*s2*s3);
our_commute  = simplify(s1*s3 - s3*s1);
law_braid12  = simplify(L_s1*L_s2*L_s1 - L_s2*L_s1*L_s2);
law_braid23  = simplify(L_s2*L_s3*L_s2 - L_s3*L_s2*L_s3);
law_commute  = simplify(L_s1*L_s3 - L_s3*L_s1);

fprintf('  Our sigma_1 sigma_2 sigma_1 = sigma_2 sigma_1 sigma_2: %s\n', ...
    mat2str(all(all(our_braid12 == 0))));
fprintf('  Our sigma_2 sigma_3 sigma_2 = sigma_3 sigma_2 sigma_3: %s\n', ...
    mat2str(all(all(our_braid23 == 0))));
fprintf('  Our sigma_1 sigma_3 = sigma_3 sigma_1:                 %s\n', ...
    mat2str(all(all(our_commute == 0))));
fprintf('  Law sigma_1 sigma_2 sigma_1 = sigma_2 sigma_1 sigma_2: %s\n', ...
    mat2str(all(all(law_braid12 == 0))));
fprintf('  Law sigma_2 sigma_3 sigma_2 = sigma_3 sigma_2 sigma_3: %s\n', ...
    mat2str(all(all(law_braid23 == 0))));
fprintf('  Law sigma_1 sigma_3 = sigma_3 sigma_1:                 %s\n', ...
    mat2str(all(all(law_commute == 0))));

fprintf('\n=== Simultaneous conjugacy verified for B_4. ===\n');
